clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/KafkaConsumerService.java

```diff
@@ -237,8 +239,10 @@ public void unsubscribeAll(PubSubTopic versionTopic) {
   public void unSubscribe(PubSubTopic versionTopic, PubSubTopicPartition pubSubTopicPartition) {
     PubSubConsumerAdapter consumer = getConsumerAssignedToVersionTopicPartition(versionTopic, pubSubTopicPartition);
     if (consumer != null) {
-      consumer.unSubscribe(pubSubTopicPartition);
-      consumerToConsumptionTask.get(consumer).removeDataReceiver(pubSubTopicPartition);
+      synchronized (consumer) {
```

Comment:
Could you also check  function `batchUnsubscribe` in this file? feel like it also needs a similar fix also need a similar fix.

Reply:
As mentioned in test, `batchUnsubscribe` is only for batch-only stores, it will not unsubscribe real-time topic partition. This issue is mainly related to real-time topic partition.

Reply:
If this assumption holds for future usage, then let's make it clear in the comments that don't use `batchUnsubscribe` for unsubscribing real time topic partition. Otherwise, later who forgets about such requirements would spend another round of time and energy to debug a similar issue.